### PR TITLE
FreeIPA inventory (ipalib 4.6.2): avoid exception

### PR DESCRIPTION
**Pull request by pilou-**
##### SUMMARY

FreeIPA inventory:
- Workaround [7345](https://pagure.io/freeipa/issue/7345) FreeIPA 4.6.2 issue: require some parameters in order to avoid an `AttributeError` exception
- mentions `IPA_CONFDIR` environment variable.

Exception:
```
$ ./freeipa.py --list
Traceback (most recent call last):
  File "./ansible/contrib/inventory/freeipa.py", line 95, in <module>
    api = initialize()
  File "./ansible/contrib/inventory/freeipa.py", line 18, in initialize
    api.finalize()
  File "local/lib/python2.7/site-packages/ipalib/plugable.py", line 738, in finalize
    self.__do_if_not_done('load_plugins')
  File "local/lib/python2.7/site-packages/ipalib/plugable.py", line 425, in __do_if_not_done
    getattr(self, name)()
  File "local/lib/python2.7/site-packages/ipalib/plugable.py", line 618, in load_plugins
    for package in self.packages:
  File "local/lib/python2.7/site-packages/ipalib/__init__.py", line 949, in packages
    ipaclient.remote_plugins.get_package(self),
  File "local/lib/python2.7/site-packages/ipaclient/remote_plugins/__init__.py", line 120, in get_package
    server_info = ServerInfo(api)
  File "local/lib/python2.7/site-packages/ipaclient/remote_plugins/__init__.py", line 26, in __init__
    hostname = DNSName(api.env.server).ToASCII()
AttributeError: 'Env' object has no attribute 'server
```

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
contrib/inventory/freeipa.py

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 18ea816a0e) last updated 2017/12/21 19:38:37 (GMT -400)
```

###### WITHOUT THIS PULL REQUEST

Without this pull request, using 4.6.2, without any configuration:
```
$ pip list|grep ipa
ipaclient (4.6.2)
ipalib (4.6.2)
ipaplatform (4.6.2)
ipapython (4.6.2)

$ ./freeipa.py --list
Traceback (most recent call last):
  File "./ansible/contrib/inventory/freeipa.py", line 95, in <module>
    api = initialize()
[...]
  File "local/lib/python2.7/site-packages/ipaclient/remote_plugins/__init__.py", line 26, in __init__
    hostname = DNSName(api.env.server).ToASCII()
AttributeError: 'Env' object has no attribute 'server'
```

Without this pull request, using 4.6.2, with an empty configuration directory:
```
$ IPA_CONFDIR="$(mktemp -d)" ./freeipa.py --list
Traceback (most recent call last):
  File "./ansible/contrib/inventory/freeipa.py", line 95, in <module>
    api = initialize()
[...]
  File "local/lib/python2.7/site-packages/ipaclient/remote_plugins/__init__.py", line 26, in __init__
    hostname = DNSName(api.env.server).ToASCII()
AttributeError: 'Env' object has no attribute 'server'
```

Without this PR, using 4.6.2. with a configuration:
```
$ cat ~/default.conf
[global]
domain=ipa.test
xmlrpc_uri=https://server.ipa.test/ipa/xml

$ IPA_CONFDIR=~ ./freeipa.py --list
{
 "_meta": {
  "hostvars": {
   "replica.ipa.test": {}, 
   "server.ipa.test": {}
  }
 }, 
 "debian-servers": {
  "hosts": [
   "server.ipa.test"
  ]
 }, 
 "ipaservers": {
  "hosts": [
   "server.ipa.test", 
   "replica.ipa.test"
  ]
 }
}
```

###### WITH THIS PULL REQUEST

With this PR, using 4.6.2, without any IPA configuration:
```
$ ./freeipa.py --list
WARNING: IPA configuration directory (/etc/ipa) is missing. Environment variable IPA_CONFDIR could be used to override default path.
ERROR: ('jsonrpc_uri' or 'xmlrpc_uri') or 'domain' are not defined in '[global]' section of '/etc/ipa/cli.conf' nor in '/etc/ipa/default.conf'.
```

With this PR, using 4.6.2, with an empty configuration directory:
```
$ IPA_CONFDIR="$(mktemp -d)" ./freeipa.py --list
ERROR: ('jsonrpc_uri' or 'xmlrpc_uri') or 'domain' are not defined in '[global]' section of '/tmp/tmp.7wWKDelszq/cli.conf' nor in '/tmp/tmp.7wWKDelszq/default.conf'.
```

With this PR, using 4.6.2, with a configuration (nothing changed):
```
$ cat ~/default.conf
[global]
domain=ipa.test
xmlrpc_uri=https://server.ipa.test/ipa/xml

$ IPA_CONFDIR=~ ./freeipa.py --list
{
 "_meta": {
  "hostvars": {
   "replica.ipa.test": {}, 
   "server.ipa.test": {}
  }
 }, 
 "debian-servers": {
  "hosts": [
   "server.ipa.test"
  ]
 }, 
 "ipaservers": {
  "hosts": [
   "server.ipa.test", 
   "replica.ipa.test"
  ]
 }
}
```

With 4.5.X, the only change is when `IPA_CONFDIR` isn't set and default value (`/etc/ipa`) doesn't exist:
```
$ ./freeipa.py --list
WARNING: IPA configuration directory (/etc/ipa) is missing. Environment variable IPA_CONFDIR could be used to override default path.
Traceback (most recent call last):
  File "./freeipa.py", line 113, in <module>
    api = initialize()
[...]
    raise NetworkError(uri=server, error=str(e))
ipalib.errors.NetworkError: cannot connect to 'http://localhost:8888/ipa/json': [Errno 111] Connection refused
```

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @Nosmoht @fxfitz

File: contrib/inventory/freeipa.py [hunk 1]
```diff
@@ -3,18 +3,41 @@
...
+
+    if LooseVersion(IPA_VERSION) == LooseVersion('4.6.2'):
+        # With ipalib < 4.6.0 'server' and 'domain' have default values
+        # ('localhost:8888', 'example.com'), newer versions don't and
+        # DNS autodiscovery is broken, then one of jsonrpc_uri / xmlrpc_uri is
+        # required.
+        # ipalib 4.6.0 is unusable (https://pagure.io/freeipa/issue/7132)
+        # that's why 4.6.2 is explicitely tested.
+        if 'server' not in api.env or 'domain' not in api.env:
+            print("ERROR: ('jsonrpc_uri' or 'xmlrpc_uri') or 'domain' are not "
```
Comment (Akasurde) [inline contrib/inventory/freeipa.py]: print can be replaced using sys.exit 

Except this looks good to me.

**Comment by Akasurde** [comment]
@pilou- Could you please rebase ? 

cc @fxfitz @flo-renaud

**Review by fxfitz — APPROVED** [review]
LGTM

File: contrib/inventory/freeipa.py (hunk 1 above)
Comment (pilou-) [reply contrib/inventory/freeipa.py]: Done

**Comment by pilou-** [comment]
PR updated:
- comment has been addressed
- PR has been rebased

**Comment by pilou-** [comment]
ready_for_review

**Comment by pilou-** [comment]
ready_for_review

File: contrib/inventory/freeipa.py [hunk 2]
```diff
@@ -3,19 +3,41 @@
...
     '''
 
     api.bootstrap(context='cli')
+
+    if not os.path.isdir(api.env.confdir):
+        print("WARNING: IPA configuration directory (%s) is missing. "
+              "Environment variable IPA_CONFDIR could be used to override "
+              "default path." % api.env.confdir)
+
+    if LooseVersion(IPA_VERSION) == LooseVersion('4.6.2'):
```
Comment (flo-renaud) [inline contrib/inventory/freeipa.py]: Hi,

ipa 4.6.3 has already been released and would fail this test. Shouldn't you use >= rather than == ?

Comment (pilou-) [reply contrib/inventory/freeipa.py]: Done

**Comment by flo-renaud** [comment]
LGTM

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by Akasurde** [comment]
@fxfitz @flo-renaud Thanks for reviews. @pilou- Thanks for contribution.

**Pull request merged**